1006: GroupCollective setup races with continuation registration code

cz4rs opened the pull request:
When a collective registration is delayed on one node and it starts sending messages before the continuations are registered, the continuation runs but doesn't have the proper state to make progress.

Solution:
Move `makeCollectiveContinuation` call after continuation registrations.
This must be *after* all the other continuation registrations because it triggers buffered work that is waiting on the meta-continuation to be registered which checks for message counts and then dispatches to the secondary continuation.

A few more details: The call`makeCollectiveContinuation(group)`  takes the new `group` ID that is mid-construction as an argument and creates the "meta-continuation" for it, which basically says that the message dispatchers are ready for that group so messages that are held back can not be delivered. If that happens before they are all actually registered (which was the ordering previously) we get bogus data for the registered handlers.

fixes #1006

cz4rs commented:
a635b21 reproduces the issue reliably in tutorial, but should be replaced with an actual (extended) unit test

cz4rs commented on tests/unit/termination/test_term_dep_send_chain.cc:
```diff
@@ -671,20 +671,17 @@ TEST_P(TestTermDepSendChain, test_term_dep_send_chain_merge) {
   vt::theConfig()->vt_term_rooted_use_ds = use_ds;
 
   auto obj_a = std::make_unique<MergeObjGroup>();
```
reverted a workaround introduced to avoid this bug

cz4rs commented on tests/unit/group/test_group.extended.cc:
```diff
@@ -0,0 +1,100 @@
+
+  if (this_node == 0) {
+    int val = 10;
+    std::this_thread::sleep_for(std::chrono::seconds(1));
```
FIXME: one second is not enough (?) too many builds are passing

cz4rs commented:
<details><summary>macosx hang detected log</summary>
<p>

```
  2/581 Test   #2: vt_example:tutorial_2 ..............................................................***Failed    1.77 sec
vt: Runtime Initializing: mode: single-thread per rank
vt: Program: tutorial (./tutorial)
vt: Running on: 2 nodes
vt: Machine Hostname: Mac-1617897483494.local
vt: MPI Version: 3.1
vt: MPI Max tag: 268435455
vt: Build SHA: 36e7b782d617cf413a2cc77b93f4fff44c50dd8c
vt: Build Ref: 
vt: Description: remotes/pull/1325/merge-0-g36e7b782d6 
vt: Compile-time Features Enabled:
vt: 	Tracing Projections
vt: 	Load Balancing for Collections
vt: 	Message priorities
vt: 	Memory Pooling
vt: 	Performance analysis diagnostics
vt: Runtime Configuration:
vt: 	Option: flag --vt_sched_num_progress on: Running MPI progress 2 times each invocation
vt: 	Option: flag --vt_sched_progress_han on: Running MPI progress function at least every 0 handler(s) executed
vt: 	Option: flag --vt_print_no_progress on: Printing warnings when progress is stalls
vt: 	Default: Printing verbose epoch graphs when hang detected, use --vt_epoch_graph_terse to disable
vt: 	Option: flag --vt_epoch_graph_on_hang on: Epoch graph output enabled if hang detected
vt: 	Default: Termination hang detection enabled by default, use --vt_no_detect_hang to disable
vt: 	Option: flag --vt_hang_detect on: Printing stall warning every 1024 tree traversals 
vt: 	Default: SIGINT signal handling enabled by default, use --vt_no_SIGINT to disable
vt: 	Default: SIGSEGV signal handling enabled by default, use --vt_no_SIGSEGV to disable
vt: 	Default: std::terminate signal handling enabled by default, use --vt_no_terminate to disable
vt: 	Default: Color output enabled, use --vt_no_color to disable
vt: 	Default: Stack dumps enabled by default, use --vt_no_stack to disable
vt: 	Option: flag --vt_memory_reporters on: Memory usage checker precedence: mstats,machinfo,selfstat,selfstatm,sbrk,mallinfo,getrusage,ps
vt: 	Working memory reporters: mstats,machinfo,getrusage
vt: 	Initial memory usage: mstats=4.55707 machinfo=13.0938 getrusage=13.1055 (MiB)
vt: 	Default: Memory usage printing disabled, use --vt_print_memory_each_phase to enable
vt: 	Default: Diagnostics are disabled by default, use --vt_diag_enable to enable
vt: 	Option: flag --vt_max_mpi_send_size on: Splitting messages greater than 1 GiB
vt: 	Option: flag --vt_debug_level on: Debug print mode set to terse
vt: Pass-through Arguments:
vt: 	None. All arguments handled.
this_node=0, num_nodes=2
msgHandlerA: triggered on node=1
msgHandlerX: triggered on node=1
msgSerialA: triggered on node=1
getCallbackHandler: triggered on node=1
msgHandlerX: triggered on node=0
getCallbackHandler: triggered on node=1
msgHandlerB: triggered on node=0
getCallbackHandler: triggered on node=1
Group is created: id=2
0: triggering active message callback
getCallbackHandler: triggered on node=1
0: triggering active message callback bcast
1: triggering active message callback bcast
0: triggering context callback
msgHandlerGroupA: triggered on node=1
0: triggering void function callback
Group is created: id=7ffff
msgHandlerGroupB: triggered on node=0
Group is created: id=7ffff
reduction value=100
0:hangCheckHandler
1:hangCheckHandler
file=epoch_graph.0.dot
file=epoch_graph.1.dot
file desc=0x7fff8fe59b60
file desc=0x7fff8fe59b60
file=epoch_graph.global.dot
file desc=0x7fff8fe59b60
file=epoch_graph.0.dot
file=epoch_graph.1.dot
file desc=0x7fff8fe59b60
file desc=0x7fff8fe59b60
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] ------------------------------------------- Runtime Error: System Aborting! --------------------------------------------
vt: [0] ------------------------------------------------ Fatal Error on Node 0 -------------------------------------------------
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] 
vt: [0]              Reason: Detected hang indicating no further progress is possible
vt: [0]                Type: vtAbort() Invoked
vt: [0]                Node: 0
vt: [0]           Num Nodes: 2
vt: [0]                File: /Users/runner/work/vt/vt/src/vt/termination/termination.cc
vt: [0]                Line: 481
vt: [0]            Function: propagateEpoch
vt: [0]                Code: 1
vt: [0]           Build SHA: 36e7b782d617cf413a2cc77b93f4fff44c50dd8c
vt: [0]           Build Ref: 
vt: [0]         Description: remotes/pull/1325/merge-0-g36e7b782d6
vt: [0]            GIT Repo: clean
vt: [0]            Hostname: Mac-1617897483494.local
vt: [0] 
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] -------------------------------------------- Dump Stack Backtrace on Node 0 --------------------------------------------
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] 0   18  0x10dfa6fc0   vt::debug::stack::dumpStack(int) + 64
vt: [0] 1   18  0x10e153d5b   vt::runtime::Runtime::output(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, bool, bool, bool) + 2987
vt: [0] 2   18  0x10dff136b   vt::CollectiveAnyOps<(vt::runtime::eRuntimeInstance)0>::output(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, bool, bool, bool, bool) + 107
vt: [0] 3   18  0x10dfe9466   vt::output(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, bool, bool, bool, bool) + 70
vt: [0] 4   18  0x10de401f8   std::__1::enable_if<(std::tuple_size<std::__1::tuple<> >::value) == (0), void>::type vt::error::displayLoc<>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::tuple<>&&) + 168
vt: [0] 5   18  0x10e09b903   vt::term::TerminationDetector::propagateEpoch(vt::term::TermState&) + 1491
vt: [0] 6   18  0x10e4d9d72   vt::runnable::RunnableNew::run() + 850
vt: [0] 7   18  0x10e4ee0ab   vt::sched::BaseUnit::execute() + 27
vt: [0] 8   18  0x10e4f04f9   vt::sched::Scheduler::runSchedulerOnceImpl(bool) + 585
vt: [0] 9   18  0x10e4f0dba   vt::sched::Scheduler::runSchedulerWhile(std::__1::function<bool ()>) + 106
vt: [0] 10  18  0x10e09b635   vt::term::TerminationDetector::propagateEpoch(vt::term::TermState&) + 773
vt: [0] 11  18  0x10e4d9d72   vt::runnable::RunnableNew::run() + 850
vt: [0] 12  18  0x10e4ee0ab   vt::sched::BaseUnit::execute() + 27
vt: [0] 13  18  0x10e4f04f9   vt::sched::Scheduler::runSchedulerOnceImpl(bool) + 585
vt: [0] 14  18  0x10e4f0dba   vt::sched::Scheduler::runSchedulerWhile(std::__1::function<bool ()>) + 106
vt: [0] 15  18  0x10e4f126f   vt::runSchedulerThrough(unsigned long long) + 111
vt: [0] 16  18  0x10e02b3b1   vt::group::InfoColl::setupCollective() + 1905
vt: [0] 17  18  0x10e428744   vt::group::Info::setup() + 1124
vt: [0] 18  18  0x10e42a445   vt::group::GroupManager::initializeLocalGroupCollective(unsigned long long, bool, std::__1::function<void ()>, bool, bool) + 213
vt: [0] 19  18  0x10e42a1e3   vt::group::GroupManager::newCollectiveGroup(bool, bool, std::__1::function<void (unsigned long long)>, bool) + 451
vt: [0] 20  18  0x10e429fb0   vt::group::GroupManager::newGroupCollective(bool, std::__1::function<void (unsigned long long)>, bool) + 112
vt: [0] 21  18  0x10de8eb96   unsigned long long vt::vrt::collection::CollectionManager::createGroupCollection<vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >(unsigned long long const&, bool) + 118
vt: [0] 22  18  0x10def91f7   void vt::vrt::collection::CollectionManager::groupConstruction<vt::tutorial::MyCol>(unsigned long long const&, bool) + 103
vt: [0] 23  18  0x10def89a7   void vt::vrt::collection::CollectionManager::distConstruct<vt::vrt::collection::CollectionCreateMsg<vt::vrt::collection::CollectionInfo<vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >, std::__1::tuple<>, vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> > >(vt::vrt::collection::CollectionCreateMsg<vt::vrt::collection::CollectionInfo<vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >, std::__1::tuple<>, vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >*) + 583
vt: [0] 24  18  0x10df8c4ad   vt::vrt::collection::CollectionProxy<vt::tutorial::MyCol, vt::tutorial::MyCol::IndexType> vt::vrt::collection::CollectionManager::constructMap<vt::tutorial::MyCol>(vt::tutorial::MyCol::IndexType, long long) + 637
vt: [0] 25  18  0x10de3c191   main + 5665
vt: [0] 26  18  0x7fff698aacc9 start + 1
vt: [0] 27  18  0x1           27  ???                                 0x0000000000000001 0x0 + 1 + 0
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] ------------------------------------------- Runtime Error: System Aborting! --------------------------------------------
vt: [0] ------------------------------------------------ Fatal Error on Node 0 -------------------------------------------------
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] 
vt: [0] Message: Assertion Failed
vt: [0] 
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] -------------------------------------------- Dump Stack Backtrace on Node 0 --------------------------------------------
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] 0   18  0x10dfa6fc0   vt::debug::stack::dumpStack(int) + 64
vt: [0] 1   18  0x10e153d5b   vt::runtime::Runtime::output(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, bool, bool, bool) + 2987
vt: [0] 2   18  0x10e15314f   vt::runtime::Runtime::abort(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) + 63
vt: [0] 3   18  0x10dff12c0   vt::CollectiveAnyOps<(vt::runtime::eRuntimeInstance)0>::abort(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) + 96
vt: [0] 4   18  0x10dfe93e6   vt::abort(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) + 38
vt: [0] 5   18  0x10dff13fb   vt::CollectiveAnyOps<(vt::runtime::eRuntimeInstance)0>::output(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, bool, bool, bool, bool) + 251
vt: [0] 6   18  0x10dfe9466   vt::output(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, bool, bool, bool, bool) + 70
vt: [0] 7   18  0x10de401f8   std::__1::enable_if<(std::tuple_size<std::__1::tuple<> >::value) == (0), void>::type vt::error::displayLoc<>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::tuple<>&&) + 168
vt: [0] 8   18  0x10e09b903   vt::term::TerminationDetector::propagateEpoch(vt::term::TermState&) + 1491
vt: [0] 9   18  0x10e4d9d72   vt::runnable::RunnableNew::run() + 850
vt: [0] 10  18  0x10e4ee0ab   vt::sched::BaseUnit::execute() + 27
vt: [0] 11  18  0x10e4f04f9   vt::sched::Scheduler::runSchedulerOnceImpl(bool) + 585
vt: [0] 12  18  0x10e4f0dba   vt::sched::Scheduler::runSchedulerWhile(std::__1::function<bool ()>) + 106
vt: [0] 13  18  0x10e09b635   vt::term::TerminationDetector::propagateEpoch(vt::term::TermState&) + 773
vt: [0] 14  18  0x10e4d9d72   vt::runnable::RunnableNew::run() + 850
vt: [0] 15  18  0x10e4ee0ab   vt::sched::BaseUnit::execute() + 27
vt: [0] 16  18  0x10e4f04f9   vt::sched::Scheduler::runSchedulerOnceImpl(bool) + 585
vt: [0] 17  18  0x10e4f0dba   vt::sched::Scheduler::runSchedulerWhile(std::__1::function<bool ()>) + 106
vt: [0] 18  18  0x10e4f126f   vt::runSchedulerThrough(unsigned long long) + 111
vt: [0] 19  18  0x10e02b3b1   vt::group::InfoColl::setupCollective() + 1905
vt: [0] 20  18  0x10e428744   vt::group::Info::setup() + 1124
vt: [0] 21  18  0x10e42a445   vt::group::GroupManager::initializeLocalGroupCollective(unsigned long long, bool, std::__1::function<void ()>, bool, bool) + 213
vt: [0] 22  18  0x10e42a1e3   vt::group::GroupManager::newCollectiveGroup(bool, bool, std::__1::function<void (unsigned long long)>, bool) + 451
vt: [0] 23  18  0x10e429fb0   vt::group::GroupManager::newGroupCollective(bool, std::__1::function<void (unsigned long long)>, bool) + 112
vt: [0] 24  18  0x10de8eb96   unsigned long long vt::vrt::collection::CollectionManager::createGroupCollection<vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >(unsigned long long const&, bool) + 118
vt: [0] 25  18  0x10def91f7   void vt::vrt::collection::CollectionManager::groupConstruction<vt::tutorial::MyCol>(unsigned long long const&, bool) + 103
vt: [0] 26  18  0x10def89a7   void vt::vrt::collection::CollectionManager::distConstruct<vt::vrt::collection::CollectionCreateMsg<vt::vrt::collection::CollectionInfo<vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >, std::__1::tuple<>, vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> > >(vt::vrt::collection::CollectionCreateMsg<vt::vrt::collection::CollectionInfo<vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >, std::__1::tuple<>, vt::tutorial::MyCol, vt::index::DenseIndexArray<int, (signed char)1> >*) + 583
vt: [0] 27  18  0x10df8c4ad   vt::vrt::collection::CollectionProxy<vt::tutorial::MyCol, vt::tutorial::MyCol::IndexType> vt::vrt::collection::CollectionManager::constructMap<vt::tutorial::MyCol>(vt::tutorial::MyCol::IndexType, long long) + 637
vt: [0] 28  18  0x10de3c191   main + 5665
vt: [0] 29  18  0x7fff698aacc9 start + 1
vt: [0] 30  18  0x1           30  ???                                 0x0000000000000001 0x0 + 1 + 0

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 7174 RUNNING AT Mac-1617897483494.local
=   EXIT CODE: 5
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Trace/BPT trap: 5 (signal 5)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

</p>
</details>

PhilMiller commented:
Given the subtlety of this issue and its fix, it would be very helpful to have a much fuller explanation of what was wrong and how it's fixed in the PR description.

cz4rs commented:
> Given the subtlety of this issue and its fix, it would be very helpful to have a much fuller explanation of what was wrong and how it's fixed in the PR description.

I have updated the description (with Jonathan's comment) and rebased this on top of the latest changes. For more in-depth explanation, @lifflander do you have anything else to add?

lifflander commented:
> > Given the subtlety of this issue and its fix, it would be very helpful to have a much fuller explanation of what was wrong and how it's fixed in the PR description.
> 
> I have updated the description (with Jonathan's comment) and rebased this on top of the latest changes. For more in-depth explanation, @lifflander do you have anything else to add?

Looks good to me!

lifflander commented:
I'm going to approve this but we might want a secondary approval.,

Also, this bug **applies to release** and should be pushed out with the next verison. I'm marking the issue as such. We might need to cut out the `vt_debug_print` fix.,
